packages/flutter/lib/src/widgets/asset_vendor.dart

```diff
@@ -0,0 +1,200 @@
+      return candidates[value];
+    double lower = candidates.lastKeyBefore(value);
+    double upper = candidates.firstKeyAfter(value);
+    if (lower == null)
+      return candidates[upper];
+    if (upper == null)
+      return candidates[lower];
+    if (value > (lower + upper) / 2)
+      return candidates[upper];
+    else
+      return candidates[lower];
+  }
+
+  String chooseVariant(String main, List<String> candidates) {
+    SplayTreeMap<double, String> mapping = _buildMapping(candidates);
+    // We assume the main asset is designed for a device pixel ratio of 1.0
+    mapping[1.0] = main;
+    return _findNearest(mapping, devicePixelRatio);
+  }
+}
+
+/// Establishes an asset resolution strategy for its descendants.
+///
+/// Given a main asset and a set of variants, AssetVendor chooses the most
+/// appropriate asset for the current context. The current asset resolution
+/// strategy knows how to find the asset most closely matching the current
+/// device pixel ratio, as given by [MediaQueryData].
+///
+/// Main assets are presumed to match a nominal pixel ratio of 1.0. To specify
+/// assets targeting different pixel ratios, place the variant assets in
+/// the application bundle under subdirectories named in the form "Nx", where
+/// N is the nominal device pixel ratio for that asset.
+///
+/// For example, suppose an application wants to use an icon named
+/// "heart.png". This icon has representations at 1.0 (the main icon), as well
+/// as 1.5 and 2.0 pixel ratios (variants). The asset bundle should then contain
+/// the following assets:
+///
+/// heart.png
+/// 1.5x/heart.png
+/// 2.0x/heart.png
+///
+/// On a device with a 1.0 device pixel ratio, the image chosen would be
+/// heart.png; on a device with a 1.3 device pixel ratio, the image chosen
+/// would be 1.5x/heart.png.
+///
+/// The directory level of the asset does not matter as long as the variants are
+/// at the equivalent level; that is, the following is also a valid bundle
+/// structure:
+///
+/// icons/heart.png
+/// icons/1.5x/heart.png
+/// icons/2.0x/heart.png
+class AssetVendor extends StatefulComponent {
+  AssetVendor({
+    Key key,
+    this.bundle,
+    this.devicePixelRatio,
+    this.child
+  }) : super(key: key);
+
+  final AssetBundle bundle;
+  final double devicePixelRatio;
+  final Widget child;
+
+  _AssetVendorState createState() => new _AssetVendorState();
+}
+
+class _AssetVendorState extends State<AssetVendor> {
+
+  _ResolvingAssetBundle _bundle;
+
+  void initState() {
+    super.initState();
+    _bundle = new _ResolvingAssetBundle(
+      bundle: config.bundle,
+      resolver: new _ResolutionAwareAssetResolver(
+        bundle: config.bundle,
+        devicePixelRatio: config.devicePixelRatio
+      )
+    );
+  }
+
+  void didUpdateConfig(AssetVendor oldConfig) {
+    if (config.bundle != oldConfig.bundle ||
+        config.devicePixelRatio != oldConfig.devicePixelRatio) {
+      setState(() {
```

Comment:
No need to use setState inside didUpdateConfig.  The system already knows that a new config means a new state.